### ISSUE TITLE
Windows configuration not restored after unzoom

### ISSUE THREAD

**Issue by tmalsburg** [emacsorphanage/zoom-window#19]
Cool package!  One issue, though: Say I have two windows each showing one buffer. Now, when I
- `zoom-window-zoom`,
- then `zoom-window-next`,
- and then unzoom via `zoom-window-zoom`,

I do not get the original window/buffer configuration back. Instead, I see two windows but they both show the same buffer.  Perhaps this behavior is actually intended, but I think it would be nicer and more predictable if we'd get back to the original configuration after unzooming.

**Comment by syohex** [comment]
I'm not sure that a following patch fixes this issue.

If the buffer which is displayed before unzoom is in original `window-list`, then zoom-window.el does not switch buffer.

```diff
diff --git a/zoom-window.el b/zoom-window.el
index 7522eb6..eb4dd59 100644
--- a/zoom-window.el
+++ b/zoom-window.el
@@ -312,6 +312,13 @@ PERSP: the perspective to be killed."
   (goto-char (point-min))
   (forward-line (1- line)))
 
+(defun zoom-window--restore-buffer (prev-buf)
+  (let ((buf (cl-loop for win in (window-list)
+                      when (eq prev-buf (window-buffer win))
+                      return prev-buf)))
+    (unless buf
+      (switch-to-buffer prev-buf))))
+
 (defun zoom-window--do-unzoom ()
   "Not documented."
   (let ((current-line (line-number-at-pos))
@@ -319,8 +326,7 @@ PERSP: the perspective to be killed."
         (current-buf (current-buffer)))
     (zoom-window--restore-mode-line-face)
     (zoom-window--restore-window-configuration)
-    (unless (string= (buffer-name current-buf) (buffer-name))
-      (switch-to-buffer current-buf))
+    (zoom-window--restore-buffer current-buf)
     (zoom-window--goto-line current-line)
     (move-to-column current-column)))
 ```

**Comment by tmalsburg** [comment]
Thanks for the patch. Will try it.  I'm not the biggest expert for managing windows and buffers but wouldn't it be enough to save the window configuration including buffers via `current-window-configuration` when the user zooms in and to restore that configuration with `set-window-configuration` when they zoom out?

**Comment by tmalsburg** [comment]
p.s: I see the current code is using those functions. But I don't understand why it's not working as intended.  The documentation of `current-window-configuration` says:

> If FRAME is nil or omitted, use the selected frame.
> This describes the number of windows, their sizes and current buffers,
> and for each displayed buffer, where display starts, and the position of
> point.

That should actually do the job.

**Comment by tmalsburg** [comment]
I think the problem is that `zoom-window--do-unzoom` switches to the currently edited buffer (the one we see after `zoom-window-next`) even where there already is a window displaying the buffer.

The following patch works for me:

``` diff
diff --git a/zoom-window.el b/zoom-window.el
index 7522eb6..6ffc312 100644
--- a/zoom-window.el
+++ b/zoom-window.el
@@ -319,8 +319,7 @@ PERSP: the perspective to be killed."
         (current-buf (current-buffer)))
     (zoom-window--restore-mode-line-face)
     (zoom-window--restore-window-configuration)
-    (unless (string= (buffer-name current-buf) (buffer-name))
-      (switch-to-buffer current-buf))
+    (pop-to-buffer current-buf)
     (zoom-window--goto-line current-line)
     (move-to-column current-column)))
 
```
Do I understand correctly that this package doesn't have a maintainer at this time?  How would be go about getting a fix merged?  Not familiar with emacsorphanage, sorry.

**Comment by jcs090218** [comment]
Hi, I am the current maintainer for this package. If you wish something to merge please open the PR here then I will review it! ;)

**Comment by tmalsburg** [comment]
Thank you, @jcs090218. I just saw that @syohex wrote this package originally, so he's in a better position to decide which patch is the best solution. But I'd be happy to make a PR if that helps.

**Comment by jcs090218** [comment]
I respect to the community; if anyone chime in and find out the solution is useful I will merge to benefit Emacs community.

I think @syohex is trying to resolve your issue here; so I guess if your patch works then it should be the one that merged. Unless anyone has the objection to this patch! 😅

**Comment by tmalsburg** [comment]
Oh, sorry I didn't want to offend anyone.  And of course you, @jcs090218, as the maintainer can make the decision as you see fit. (As I said, I don't understand yet how emacsorphanage works.)  The reason why I hesitated to make a PR is that I don't know what the intended behavior is.  My patch behaves the way I'd prefer it, but I saw that @syohex solution is a bit more complex, so I it's possible that I'm overlooking something.  Either way, I've created a PR: #20.

Thanks both for your input.  Much appreciated.

**Comment by jcs090218** [comment]
> Oh, sorry I didn't want to offend anyone. 

No, I don't see anyone offend anyone here! Only friendly communication here. ;)

> My patch behaves the way I'd prefer it, but I saw that @syohex solution is a bit more complex, so I it's possible that I'm overlooking something.

emacsorphange is where people maintain packages that the original author is no longer interested maintaining their own packages. I am maintaining this package because I feel like I am able to help! I think anyone can step in to volunteer to maintain any package that you are interested!

> The reason why I hesitated to make a PR is that I don't know what the intended behavior is. My patch behaves the way I'd prefer it, but I saw that @syohex solution is a bit more complex, so I it's possible that I'm overlooking something.

Hmm... From my understanding. This issue is more like an `enhancement`? So I guess there are no intended behavior here? 😕 And I appreciate @syohex is being very responsible when there is issue from the community!

> Either way, I've created a PR: #20.

Thanks for the PR! I will wait for a day or two before I merge this PR. Just to ensure no one is objecting this enhancement!

**Comment by tmalsburg** [comment]
> emacsorphange is where people maintain packages that the original author is no longer interested maintaining their own packages. I am maintaining this package because I feel like I am able to help!

Wow, I see there are a bunch of packages in emacsorphanage that I'm actually using. Very grateful for your contribution to this project and for helping to keep these packages alive!

> Thanks for the PR! I will wait for a day or two before I merge this PR. Just to ensure no one is objecting this enhancement!

Makes sense. :)

**Comment by jcs090218** [comment]
Merged PR #20. Close this issue now! Thanks! ;)
